virt_lightning/shell.py

```diff
@@ -72,56 +101,88 @@ def myDomainEventAgentLifecycleCallback(conn, dom, state, reason, opaque):
+    async def deploy():
+        futures = []
+        for host in virt_lightning_yaml:
+            futures.append(
+                loop.run_in_executor(
+                    pool, _start_domain, hv, host, context, configuration
+                )
+            )
 
-        if hv.get_domain_by_name(host["name"]):
-            logger.info("Skipping {name}, already here.".format(**host))
-            return
+        domain_reachable_futures = []
+        for f in futures:
+            await f
+            domain = f.result()
+            if domain:
+                domain_reachable_futures.append(domain.reachable())
+        logger.info("%s ok Waiting...", symbols.HOURGLASS.value)
 
-        # Unfortunatly, i can't decode that symbol
-        # that symbol more well add to check encoding block
-        logger.info(
-            "{lightning} {name} ".format(lightning=symbols.LIGHTNING.value, **host)
-        )
-        domain = hv.create_domain(name=host["name"], distro=host["distro"])
-        domain.context = context
-        domain.groups = host.get("groups", [])
-        domain.load_ssh_key_file(configuration.ssh_key_file)
-        if host["distro"].startswith("esxi"):
-            domain.username = "root"
-        else:
-            domain.username = configuration.username
-        domain.root_password = host.get("root_password", configuration.root_password)
-
-        domain.vcpus(host.get("vcpus"))
-        domain.memory(host.get("memory", 768))
-        root_disk_path = hv.create_disk(
-            name=host["name"],
-            backing_on=host["distro"],
-            # NOTE: Use to be 15GB, but FreeBSD root FS is 31G large
-            size=host.get("root_disk_size", 32),
-        )
-        domain.add_root_disk(root_disk_path)
-        domain.attachNetwork(configuration.network_name)
-        domain.ipv4 = hv.get_free_ipv4()
-        domain.add_swap_disk(hv.create_disk(host["name"] + "-swap", size=1))
-        hv.start(domain)
-        return domain
+        await asyncio.gather(*domain_reachable_futures)
 
-    pool = ThreadPoolExecutor(max_workers=10)
     loop.run_until_complete(deploy())
     logger.info("%s You are all set", symbols.THUMBS_UP.value)
 
 
+def start(configuration, **kwargs):
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_network(configuration.network_name, configuration.network_cidr)
+    hv.init_storage_pool(configuration.storage_pool)
+    host = {"distro": kwargs["distro"]}
+    context = "default"
+    domain = _start_domain(hv, host, context, configuration)
+    if not domain:
+        return
+    import time
+
+    time.sleep(4)
+    stream = conn.newStream(libvirt.VIR_STREAM_NONBLOCK)
+    console = domain.dom.openConsole(None, stream, 0)
+    loop = asyncio.get_event_loop()
+    import libvirtaio
+
+    libvirtaio.virEventRegisterAsyncIOImpl(loop=loop)
+
+    def stream_callback(stream, events, _):
+        line = stream.recv(1024).decode()
```

Comment:
If it is TCP socket, it will freeze for a while until get 1024 bytes on socket to work. Looks like infinity waiting for a few cases.

Reply:
It's an unix socket. recv() will read what is available until it reachs 1024.